File: components/gmaps-flow/trunk/gmaps-web-component/google-map-polygon.js
/** 
@license
Copyright 2010-2019 Neotropic SAS <contact@neotropic.co>.

Licensed under the Apache License, Version 2.0 (the "License"); 
you may not use this file except in compliance with the License. 
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software 
distributed under the License is distributed on an "AS IS" BASIS, 
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. 
See the License for the specific language governing permissions and 
limitations under the License.
*/
import {html, PolymerElement} from '@polymer/polymer/polymer-element.js';
/**
 * @author Johny Andres Ortega Ruiz {@literal <johny.ortega@kuwaiba.org>}
 */
class GoogleMapPolygon extends PolymerElement {
  static get is() {
    return 'google-map-polygon';
  }
  static get template() {
    return html`
      <style>
        :host {
          display: block;
        }
      </style>
    `;
  }
  static get properties() {
    return {
      clickable: {
        type: Boolean,
        value: true,
        observer: '_clickableChanged'
      },
      draggable: {
        type: Boolean,
        value: false,
        observer: '_draggableChanged'
      },
      editable: {
        type: Boolean,
        value: false,
        observer: '_editableChanged'
      },
      fillColor: {
        type: String,
        value: '#FF0000',
        observer: '_fillColorChanged'
      },
      fillOpacity: {
        type: Number,
        value: 0.35,
        observer: '_fillOpacityChanged'
      },
      geodesic: {
        type: Boolean,
        value: false,
        observer: '_geodesicChanged'
      },
      paths: {
        type: Array,
        observer: '_pathsChanged'
      },
      strokeColor: {
        type: String,
        value: '#FF0000',
        observer: '_strokeColorChanged'
      },
      strokeOpacity: {
        type: Number,
        value: 0.8,
        observer: '_strokeOpacityChanged'
      },
      strokeWeight: {
        type: Number,
        value: 2,
        observer: '_strokeWeightChanged'
      },
      visible: {
        type: Boolean,
        value: true,
        observer: '_visibleChanged'
      },
      zIndex: {
        type: Number,
        observer: '_zIndexChanged'
      }
    };
  }
  /**
   * @return {google.maps.MVCObject} The google.maps.Polygon
   */
  getMVCObject() {
    return this.polygon;
  }
  added(map) {
    this.polygon = new google.maps.Polygon({
      clickable: this.clickable,
      draggable: this.draggable,
      editable: this.editable,
      fillColor: this.fillColor,
      fillOpacity: this.fillOpacity,
      geodesic: this.geodesic,
      strokeColor: this.strokeColor,
      strokeOpacity: this.strokeOpacity,
      strokeWeight: this.strokeWeight,
      visible: this.visible,
    });
    this.polygon.setMap(map);
    this._setPolygonPaths(this.paths);
    this.polygon.addListener('click', event => {
      this.dispatchEvent(new CustomEvent('polygon-click'));
    });
    this.polygon.addListener('dblclick', event => {
      this.dispatchEvent(new CustomEvent('polygon-dbl-click'));
    });
    this.polygon.addListener('drag', event => {
      this.dispatchEvent(new CustomEvent('polygon-drag'));
    });
    this.polygon.addListener('dragend', event => {
      this.dispatchEvent(new CustomEvent('polygon-drag-end'));
    });
    this.polygon.addListener('dragstart', event => {
      this.dispatchEvent(new CustomEvent('polygon-drag-start'));
    });
    this.polygon.addListener('mousedown', event => {
      this.dispatchEvent(new CustomEvent('polygon-mouse-down'));
    });
    this.polygon.addListener('mousemove', event => {
      this.dispatchEvent(new CustomEvent('polygon-mouse-move'));
    });
    this.polygon.addListener('mouseout', event => {
      this.dispatchEvent(new CustomEvent('polygon-mouse-out'));
    });
    this.polygon.addListener('mouseover', event => {
      this.dispatchEvent(new CustomEvent('polygon-mouse-over'));
    });
    this.polygon.addListener('mouseup', event => {
      this.dispatchEvent(new CustomEvent('polygon-mouse-up'));
    });
    this.polygon.addListener('rightclick', event => {
      this.dispatchEvent(new CustomEvent('polygon-right-click'));
    });
  }
  _setPolygonPaths(paths) {
    this.polygon.setPaths(paths);
    
    google.maps.event.addListener(this.polygon.getPaths(), 'insert_at', index => {
      this._updatePaths();
    });
    google.maps.event.addListener(this.polygon.getPaths(), 'remove_at', index => {
      this._updatePaths();
    });
    google.maps.event.addListener(this.polygon.getPaths(), 'set_at', index => {
      this._updatePaths();
    });
    this.polygon.getPaths().forEach(path => {
      google.maps.event.addListener(path, 'insert_at', index => {
        this._updatePaths();
      });
      google.maps.event.addListener(path, 'remove_at', index => {
        this._updatePaths();
      });
      google.maps.event.addListener(path, 'set_at', index => {
        this._updatePaths();
      });
    });
  }
  _clickableChanged(newValue, oldValue) {
    if (this.polygon) {
      this.polygon.setOptions({
        clickable: newValue
      });
    }
  }
  _draggableChanged(newValue, oldValue) {
    if (this.polygon)
      this.polygon.setDraggable(newValue);
  }
  _editableChanged(newValue, oldValue) {
    if (this.polygon)
      this.polygon.setEditable(newValue);
  }
  _fillColorChanged(newValue, oldValue) {
    if (this.polygon) {
      this.polygon.setOptions({
        fillColor: newValue
      });
    }
  }
  _fillOpacityChanged(newValue, oldValue) {
    if (this.polygon) {
      this.polygon.setOptions({
        fillOpacity: newValue
      });
    }
  }
  _geodesicChanged(newValue, oldValue) {
    if (this.polygon) {
      this.polygon.setOptions({
        geodesic: newValue
      });
    }
  }
  _pathsChanged(newValue, oldValue) {
    if (this.polygon) {
      var paths = GoogleMapPolygon.getPathsArray(this.polygon.getPaths());
      if (JSON.stringify(paths) !== JSON.stringify(newValue))
        this._setPolygonPaths(newValue);
    }
  }
  _updatePaths() {
    this.paths = GoogleMapPolygon.getPathsArray(this.polygon.getPaths());
    this.dispatchEvent(new CustomEvent('polygon-paths-changed'));
  }
  static getPathsArray(paths) {
    var pathsArray = [];
    paths.forEach(path => {
      var pathArray = [];
      path.forEach(coordinate => {
        pathArray.push(coordinate.toJSON());
      });
      pathsArray.push(pathArray);
    });
    return pathsArray;
  }
  _strokeColorChanged(newValue, oldValue) {
    if (this.polygon) {
      this.polygon.setOptions({
        strokeColor: newValue
      });
    }
  }
  _strokeOpacityChanged(newValue, oldValue) {
    if (this.polygon) {
      this.polygon.setOptions({
        strokeOpacity: newValue
      });
    }
  }
  _strokeWeightChanged(newValue, oldValue) {
    if (this.polygon) {
      this.polygon.setOptions({
        strokeWeight: newValue
      });
    }
  }
  _visibleChanged(newValue, oldValue) {
    if (this.polygon)
      this.polygon.setVisible(newValue);
  }
  _zIndexChanged(newValue, oldValue) {
    if (this.polygon) {
      this.polygon.setOptions({
        zIndex: newValue
      });
    }
  }
  removed() {
    if (this.polygon)
      this.polygon.setMap(null);
  }
}
window.customElements.define(GoogleMapPolygon.is, GoogleMapPolygon);

export { GoogleMapPolygon };

File: components/gmaps-flow/version-1.0/google-map-flow/src/main/java/com/neotropic/flow/component/googlemap/GoogleMapPolygon.java
/*
 *  Copyright 2010-2020 Neotropic SAS <contact@neotropic.co>.
 *  
 *  Licensed under the Apache License, Version 2.0 (the "License"); 
 *  you may not use this file except in compliance with the License. 
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software 
 *  distributed under the License is distributed on an "AS IS" BASIS, 
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. 
 *  See the License for the specific language governing permissions and 
 *  limitations under the License.
 */
package com.neotropic.flow.component.googlemap;

import com.vaadin.flow.component.Component;
import com.vaadin.flow.component.ComponentEventListener;
import com.vaadin.flow.component.Synchronize;
import com.vaadin.flow.component.Tag;
import com.vaadin.flow.component.dependency.JsModule;
import com.vaadin.flow.shared.Registration;
import elemental.json.Json;
import elemental.json.JsonArray;
import elemental.json.JsonObject;
import elemental.json.JsonValue;
import java.util.ArrayList;
import java.util.List;

/**
 *
 * @author Johny Andres Ortega Ruiz {@literal <johny.ortega@kuwaiba.org>}
 */
@Tag("google-map-polygon")
@JsModule("./google-map-polygon.js")
public class GoogleMapPolygon extends Component {
    public GoogleMapPolygon() {
    }
    public GoogleMapPolygon(List<List<LatLng>> paths) {
        getElement().setPropertyJson(
            Constants.Property.PATHS, GoogleMapPolygon.pathsAsJsonArray(paths));
    }
    public static JsonValue pathsAsJsonArray(List<List<LatLng>> paths) {
        if (paths != null) {
            JsonArray arrayPaths = Json.createArray();
            for (int i = 0; i < paths.size(); i++) {
                JsonArray arrayPath = Json.createArray();
                List<LatLng> path = paths.get(i);
                for (int j = 0; j < path.size(); j++) {
                    JsonObject coordinate = Json.createObject();
                    coordinate.put("lat", path.get(j).getLat());
                    coordinate.put("lng", path.get(j).getLng());
                    arrayPath.set(j, coordinate);
                }
                arrayPaths.set(i, arrayPath);
            }
            return arrayPaths;
        }
        return Json.createNull();
    }
    public static List<List<LatLng>> pathsAsList(JsonArray paths) {
        if (paths != null) {
            List<List<LatLng>> listPaths = new ArrayList();
            for (int i = 0; i < paths.length(); i++) {
                List<LatLng> listPath = new ArrayList();
                JsonArray path = paths.getArray(i);
                for (int j = 0; j < path.length(); j++) {
                    JsonObject coordinate = path.getObject(j);
                    double lat = coordinate.getNumber("lat");
                    double lng = coordinate.getNumber("lng");
                    listPath.add(new LatLng(lat, lng));
                }
                listPaths.add(listPath);
            }
            return listPaths;
        }
        return null;
    }
    public boolean getEditable() {
        return getElement().getProperty(Constants.Property.EDITABLE, false);
    }
    public void setEditable(boolean editable) {
        getElement().setProperty(Constants.Property.EDITABLE, editable);
    }
    @Synchronize(property="paths", value="polygon-paths-changed")
    public List<List<LatLng>> getPaths() {
        return GoogleMapPolygon.pathsAsList(
            (JsonArray) getElement().getPropertyRaw(Constants.Property.PATHS)
        );
    }
    public void setPath(List<List<LatLng>> paths) {
        getElement().setPropertyJson(
            Constants.Property.PATHS, GoogleMapPolygon.pathsAsJsonArray(paths));
    }
    public Registration addPolygonClickListener(ComponentEventListener<GoogleMapEvent.PolygonClickEvent> listener) {
        return addListener(GoogleMapEvent.PolygonClickEvent.class, listener);
    }
    public Registration addPolygonDblClickListener(ComponentEventListener<GoogleMapEvent.PolygonDblClickEvent> listener) {
        return addListener(GoogleMapEvent.PolygonDblClickEvent.class, listener);
    }
    public Registration addPolygonMouseOutListener(ComponentEventListener<GoogleMapEvent.PolygonMouseOutEvent> listener) {
        return addListener(GoogleMapEvent.PolygonMouseOutEvent.class, listener);
    }
    public Registration addPolygonMouseOverListener(ComponentEventListener<GoogleMapEvent.PolygonMouseOverEvent> listener) {
        return addListener(GoogleMapEvent.PolygonMouseOverEvent.class, listener);
    }
    public Registration addPolygonRightClickListener(ComponentEventListener<GoogleMapEvent.PolygonRightClickEvent> listener) {
        return addListener(GoogleMapEvent.PolygonRightClickEvent.class, listener);
    }
    public Registration addPolygonPathsChangedListener(ComponentEventListener<GoogleMapEvent.PolygonPathsChangedEvent> listener) {
        return addListener(GoogleMapEvent.PolygonPathsChangedEvent.class, listener);
    }
}
